zk/conn.go

```diff
@@ -964,6 +961,25 @@ func (c *Conn) queueRequest(opcode int32, req interface{}, res interface{}, recv
 	return rq.recvChan
 }
 
+func (c *Conn) queueRequestWithTimeout(opcode int32, req interface{}, res interface{}, recvFunc func(*request, *responseHeader, error), timeout time.Duration) <-chan response {
+	rq := &request{
+		xid:        c.nextXid(),
+		opcode:     opcode,
+		pkt:        req,
+		recvStruct: res,
+		recvChan:   make(chan response, 1),
+		recvFunc:   recvFunc,
+	}
+
+	select {
+	case <-time.After(timeout):
+		rq.recvChan <- response{err: ErrConnectionClosed}
```

Comment:
in  this case we are presenting a connection closed error not based on the zookeeper state but a timeout. Is this the right way for the client to handle this?

Reply:
This is a good point.  Removed the error code sending.  If caller got a non-nil error, it means the queueing operation failed.  The receive channel thus is not valid to read.